# Intro to Data Sciecne (71253) - Problem Set 1

### Saar Pardo

## Instructions

Complete the tasks and answer each of the questions below using both text and code cells, as needed.    
Your code needs to be able to excute properly.   
Make sure to include complete answers. This can be done either using a text cell, or by formating and printing a string (using the 'format' and 'print' functions). See example below.  
  
Do **not** just dump the value of a variable onto the screen. 

### Example of desired output format

As an example, let's consider the question:  
> "How excited are you about data science from 1-10"  

and the (obvious) answer:

>10

The perfered answer formats is:

In [ ]:
excitment = 10
print("My level of excitment about data science is {ex:0}!".format(ex=excitment))

My level of excitment about data science is 10!


  
  Altenrtively, you can hard-code the answer in a text cell:   

My level of excitment about data science is 10!

This format is advised only for answers that do not contain the outputs of any code since it is prone to errors - if the code/output changes, the hard-coded text does not.

### Example of unacceptable output format

An unacceptable format is dump the value of a variable onto the screen. For example:

In [ ]:
excitment = 10
excitment

10

***

## 1. Imports and parsing

In [1]:
## Import the equired libraries for the pset here.
## It is a good habit to start with the imports you'll need in a separare code block.

import pandas as pd
import numpy as np

In [2]:
## read the data from the file "okcupid_profiles_ps1.csv" into a pandas dataframe.
## Note that the first column of in the file is already the record number (or index). 

Okcupid = pd.read_csv('okcupid_profiles_ps1.csv',index_col=0)

*** 

## 2. Overall data structure

### 2.1 How many records does the data contain?

In [3]:
len(Okcupid)

10000

* 10000 records

### 2.2 How many variables does the data contain?

In [4]:
Okcupid.shape

(10000, 21)

* 21 variables

### 2.3 What are the names of the first 5 variables?

In [5]:
Okcupid.columns.tolist()[:5]

['age', 'status', 'sex', 'orientation', 'body_type']

* the first five columns are : `age`, `status`, `sex`, `orientation`, `body_type`

***

## 3. Variable types

### 3.1 What are the types of the variables 'age', 'status', 'education', 'body_type', and 'drinks'?  
(e.g. Continous, Ordinal, etc'. Note that this is not the same as the python `dtype)

In [13]:
Okcupid[['age', 'status', 'education', 'body_type', 'drinks']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        10000 non-null  int64 
 1   status     10000 non-null  object
 2   education  8892 non-null   object
 3   body_type  9103 non-null   object
 4   drinks     9468 non-null   object
dtypes: int64(1), object(4)
memory usage: 468.8+ KB


In [18]:
Okcupid.drinks.unique()

array(['socially', 'often', 'not at all', 'rarely', nan, 'very often',
       'desperately'], dtype=object)

* The `age` type is Quantitative Discrete (integer)
* The `status` type is Qualitative Nominal (object)
* The `education` type is Qualitative Nominal (object)
* The `body_type` type is Qualitative Nominal (object)
* The `drinks` type is Qualitative Nominal (object)

### 3.2 Formulate 3 questions involving a single variable.  
(e.g. How many vegeterians are there in the data?)

In [25]:
Okcupid.orientation.unique()

array(['straight', 'bisexual', 'gay'], dtype=object)

1. how many users are defined as "bisexuals" in the data?
2. how many users are defined as "gay" in the data?
2. How many of the user data are straight?

In [36]:
b = len(Okcupid[Okcupid['orientation']=='bisexual'])
g = len(Okcupid[Okcupid['orientation']=='gay'])
s = len(Okcupid[Okcupid['orientation']=='straight'])
n =len(Okcupid)

In [37]:
print(f"{b} is the number of biosexual users in the data")
print(f"{g} is the number of gay users in the data")
print(f"{s/n:.1%} of the data has straight orientation")

451 is the number of biosexual users in the data
944 is the number of gay users in the data
86.1% of the data has straight orientation


### 3.3 Formulate 3 questions involving a pair of variables.
(e.g. How many vegeterians men are there in the data?)

In [89]:
1. how many users are defined as "bisexuals" and above 40 in the data?
2. how many users are defined as "gay" and less than 30 in the data?
2. How many of the user data are straight and obove 60?

Index(['age', 'status', 'sex', 'orientation', 'body_type', 'diet', 'drinks',
       'drugs', 'education', 'ethnicity', 'height', 'income', 'job',
       'last_online', 'location', 'offspring', 'pets', 'religion', 'sign',
       'smokes', 'speaks', 'height_cm'],
      dtype='object')

In [90]:
b1 = len(Okcupid[(Okcupid['orientation']=='bisexual') &
        (Okcupid['age']>40)])

g1 = len(Okcupid[(Okcupid['orientation']=='gay') &
        (Okcupid['age']<30)])

s1 = len(Okcupid[(Okcupid['orientation']=='straight')&
        (Okcupid['age']>60)])

n1 =len(Okcupid)

In [94]:
print(f"{b1} is the number of biosexual users that also above than 40 in the data")
print(f"{g1} is the number of gay users that also younger than 30 in the data")
print(f"{s1/n1:.1%} of the data are starights above 60")

42 is the number of biosexual users that also above than 40 in the data
508 is the number of gay users that also younger than 30 in the data
1.5% of the data are starights above 60


***

## 4. Analyzing body types

### 4.1 How many different values of the variable "body_type" are there?
(Hint: the pandas `unique` function is helpful)

In [44]:
print("there are" , f"{Okcupid.body_type.nunique()}" ,"body types in the data")

there are 12 body types in the data


### 4.2 What percentage of people report that they have an "average", "fit", or "athletic" body type?  
(sum all theree types)

In [47]:
Okcupid.body_type.unique()

array(['a little extra', 'average', 'thin', 'athletic', 'fit', nan,
       'skinny', 'curvy', 'full figured', 'jacked', 'rather not say',
       'used up', 'overweight'], dtype=object)

In [57]:
Okcupid.body_type.value_counts(dropna=False)

average           2398
fit               2127
athletic          1981
NaN                897
thin               779
curvy              664
a little extra     452
skinny             283
full figured       189
overweight          71
jacked              66
used up             51
rather not say      42
Name: body_type, dtype: int64

In [55]:
n_fit = len(Okcupid[((Okcupid['body_type']=='average')|
                        (Okcupid['body_type']=='fit')|
                        (Okcupid['body_type']=='athletic'))])
print("there are" , f"{n_fit/len(Okcupid,):.0%}" , "that belong to that group")

there are 65% that belong to that group


### (Bonus) How do you interpret the precentage of body types you computed above?  
(That is, write a couple of sentences explaining what do you think about this result and what it means.)

First of all, I didnt handle missing values so the real percentage may be even greater that 65.
I would say that according to the share of that group, most of people in Data perceive themselves as average or above average in terms of body image. There may be bias since people can give false information about themselves. 

***

## 5. Analyzing heights

### 5.1 Heights are reported in inches. Create a new column in your dataframe containing the heights in cm.  
(1 inche = 2.54cm)

In [66]:
Okcupid['height_cm'] = Okcupid['height']*2.54

In [95]:
# Okcupid.columns

### 5.2 What are the maximal and minimal heights in the data?  
(in cm)

In [68]:
maximal_h = max(Okcupid['height_cm'])
minimal_h = min(Okcupid['height_cm'])

print(f"{maximal_h} is the maximal height in the data") 
print(f"{minimal_h} is the minimal height in the data") 

241.3 is the maximal height in the data
91.44 is the minimal height in the data


### How do you interpret the results of the previous question?  
(regarding maximal and minimal height)

I would like to see the distribution in graph or percentiles because the minimum and maximum values are extreme and cause suspicion

In [83]:
Okcupid['height_cm'].describe(percentiles=[0.001,0.01,0.5,0.99,0.999])

count    10000.000000
mean       173.561248
std          9.927545
min         91.440000
0.1%       147.320000
1%         152.400000
50%        172.720000
99%        193.040000
99.9%      203.200000
max        241.300000
Name: height_cm, dtype: float64

### 5.3 How many men shorter than 170cm are there in the data? How many women?

In [96]:
m_170 = len(Okcupid[(Okcupid['height_cm']<170) & (Okcupid['sex']=='m')])
f_170 = len(Okcupid[(Okcupid['height_cm']<170) & (Okcupid['sex']=='f')])

In [97]:
print("there are" , f"{m_170} men with height under 170cm while there are", f"{f_170}","women with height under 170cm")  

there are 479 men with height under 170cm while there are 2754 women with height under 170cm


### How do you interpret the results of the previous question?  
(regarding people under 170cm)

As a man I can think of a reason for men to lie about their height on a dating app, this could be reflected in the results as only 479 of the men claimed to be under 170cm.

The average height for a man is 160.5cm (google)